transcom/mymove
# [WIP] Limit concurrent sessions & revoke sessions via admin interface

## Description

This is a rough draft with a working example of revoking sessions via the admin interface, as well as limiting concurrent sessions per app. That is, a user should be able to sign in to multiple apps (mil, admin, office) at the same time, but can only have one session per app.

## Reviewer Notes

To test session revocation:
1. Sign in at http://milmovelocal:3000/ in one browser tab
2. In another browser tab, sign in at http://adminlocal:3000/ (You might need to create yourself as an admin first via http://adminlocal:3000/devlocal-auth/login)
3. Once in the admin site, click on Search User by ID on the left sidebar
4. Enter your UUID and click the Search button. You can find your UUID in the server logs in the `middleware/request_logger.go:85` line under the key `user-id`.
5. Once your details appear in the user show page, click on Edit in the top right
6. Select "Yes" from the dropdown for "Revoke mil session"
7. Go back to the tab where you were signed into the mil app and refresh it.
Expected Result: You should be logged out.

## Setup

```sh
make db_dev_migrate
make db_test_migrate
make redis_pull
```

## Code Review Verification Steps

 [Querying the Database Safely](https://github.com/transcom/mymove/tree/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:

## References

* [MB-2329](https://dp3.atlassian.net/browse/MB-2329) for this change

## Screenshots

![session_revocation](https://user-images.githubusercontent.com/811150/82492245-ddd88700-9ab3-11ea-9437-6382604cf7ec.gif)
